src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureBlobStorage.cs

```diff
@@ -63,23 +64,26 @@ public async Task ReadStateAsync(string grainType, GrainReference grainId, IGrai
 
             try
             {
-                var blob = container.GetBlockBlobReference(blobName);
+                var blob = container.GetBlobClient(blobName);
 
                 byte[] contents;
                 try
                 {
+                    var properties = await blob.GetPropertiesAsync();
+                    grainState.ETag = properties.Value.ETag.ToString();
+
                     using (var stream = new MemoryStream())
                     {
-                        await blob.DownloadToStreamAsync(stream).ConfigureAwait(false);
+                        await blob.DownloadToAsync(stream).ConfigureAwait(false);
                         contents = stream.ToArray();
                     }
```

Comment:
`blob.DownloadToAsync` already contains the required info so this duplicate querying should be avoided:
```cs
var stream = new MemoryStream();
var response = await blob.DownloadToAsync(stream).ConfigureAwait(false);
contents = stream.ToArray();
grainState.ETag = response.Headers.ETag.ToString();